### PR TITLE
Richer stream seek interface with undo log removed

### PR DESCRIPTION
**Pull request by no2chem**
This patch reopens some of the changes introduced in #474. 
It addresses concurrency issues in the undo log. 
In particular, it introduces:
- A richer stream interface
- Elimination of the undoLog
- Elimination of the optimisticUndoLog
- Elimination of redundant fields in versionLockedObject
- Unification of the optimistic log and stream with a new ISMRStream interface
- Increased debug logging

The richer stream interface is described in #474.
Elimination of the undo logs is acheived though the richer stream interface and the
new ISMRStream implenentation. The version field is removed and the VersionLockedObject
uses the stream pointer directly to track versions. Sync/rollback have been unified
inside versionLockedObject, which has been updated to automatically perform the
appropiate rollback depending on the current version and transactional execution.

This PR is currently posted for review. The code currently lacks adequete documentation,
which will be added with additional commits.

This PR fixes the problemn highlighted in the test case in PR #488.

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **0e887f6**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #492 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/T7tApwJ3B3z57FZgCdjA7c5jMeJ7hE8p)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **f64ba44**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #492 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/DBR3QsJoQeTeK4CRMur5DdX66WLvK87F)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **f64ba44**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #492 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/1U5jZcr6AYUSsXFxv6AT3TZZHSMVqlVq)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **f1e0dfc**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #492 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/ZcEGeSrExpylmR5idVJxdSMGO85IlIGJ)

File: runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java [hunk 1]
```diff
@@ -387,10 +424,15 @@ public boolean isWriteLocked() {
...
+        } catch (InterruptedException ie) {
+            log.debug("ReadLock[{}] Timed out, retrying with write lock");
+            throw new ConcurrentModificationException();
+        }
         try {
             try {
-                return readFunction.apply(version, object);
+                return readFunction.apply(getVersionUnsafe(), this);
             } finally {
                 lock.unlock(ts);
```
Comment (slfritchie) [inline runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java]: Michael: when using Roger's MultipleNonOverlappingTest test case + adding some artificial delays in the runtime + commit f1e0dfcdc67fc09f075f55aeef07580c63dc0270 ("reduce context calls"), I see this unlock call throwing java.lang.IllegalMonitorStateException.

File: runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java [hunk 2]
```diff
@@ -281,52 +170,200 @@ public Object applyUpdateUnsafe(SMREntry entry, boolean isOptimistic) {
...
+        // and it belongs to this thread use that
+        if (optimisticStream != null &&
+                optimisticStream.isStreamForThisTransaction()) {
+            // If there are no updates, ensure we are at the right snapshot
+            if (optimisticStream.pos() == Address.NEVER_READ) {
+                final WriteSetSMRStream currentOptimisticStream =
+                        optimisticStream;
+                // It's necessary to roll back optimistic updates before
+                // doing a sync of the regular log
+                optimisticRollbackUnsafe();
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java]: why is this here? we just checked above that (a) we are the owner of the proxy, and (b) that there are not optimistic updates applied.

also, if there were any updates, it would make me a bit nervous to undo them and then redo over a different version?

File: runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java [hunk 3]
```diff
@@ -281,52 +170,200 @@ public Object applyUpdateUnsafe(SMREntry entry, boolean isOptimistic) {
...
+    }
+    /** Update the object. Ensure that you have the write lock before calling
+     * this function...
+     * @param timestamp         The timestamp to update the object to.
+     */
+    public void syncObjectUnsafe(long timestamp) {
+        // If there is an optimistic stream attached,
+        // and it belongs to this thread use that
+        if (optimisticStream != null &&
+                optimisticStream.isStreamForThisTransaction()) {
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java]: would it be cleaner to simply pass a transactionalContext as parameter? pass null if syncObject is called not from within transactional context.

(see also comment below)

File: test/src/test/resources/logback-test.xml [hunk 4]
```diff
@@ -6,9 +6,15 @@
         </encoder>
     </appender>
 
-    <logger name="org.corfudb.runtime.object" level="DEBUG"/>
+    <!-- Control logging levels for individual components here. -->
+    <logger name="org.corfudb.runtime.object" level="TRACE"/>
+    <logger name="org.corfudb.runtime.clients" level="INFO"/>
+    <logger name="org.corfudb.infrastructure" level="INFO"/>
+    <logger name="io.netty.util" level="INFO"/>
```
Comment (kspirov) [inline test/src/test/resources/logback-test.xml]: added intentionally, or forgotten debug?

File: test/src/test/java/org/corfudb/runtime/view/StreamViewTest.java [hunk 5]
```diff
@@ -109,6 +109,119 @@ public void canReadWriteFromCachedStream()
...
+                IStreamView.SearchDirection.REVERSE))
+                .isEqualTo(A_GLOBAL);
+        // Should find entry "e"
+        assertThat(svA.find(E_GLOBAL,
+                IStreamView.SearchDirection.FORWARD_INCLUSIVE))
+                .isEqualTo(E_GLOBAL);
+        // Should find entry "c"
+        assertThat(svA.find(C_GLOBAL,
+                IStreamView.SearchDirection.REVERSE_INCLUSIVE))
+                .isEqualTo(C_GLOBAL);
```
Comment (kspirov) [inline test/src/test/java/org/corfudb/runtime/view/StreamViewTest.java]: Add a  bit more (and use D_GLOBAL :)

        // From existing to existing:
        // Should find entry "b"
        assertThat(svB.find(D_GLOBAL,
                IStreamView.SearchDirection.REVERSE))
                .isEqualTo(B_GLOBAL);
        // Should find entry "d"
        assertThat(svB.find(B_GLOBAL,
                IStreamView.SearchDirection.FORWARD))
                .isEqualTo(D_GLOBAL);

        // Bounds:
        assertThat(svB.find(D_GLOBAL,
                IStreamView.SearchDirection.FORWARD))
                .isEqualTo(Address.NEVER_READ);
    
BTW why is Address.NEVER_READ, shouldn't it be Addres.NOT_FOUND?

File: test/src/test/java/org/corfudb/runtime/view/AbstractViewTest.java [hunk 6]
```diff
@@ -108,8 +108,8 @@ private IClientRouter getRouterFunction(CorfuRuntime runtime, String endpoint) {
     @Before
     public void resetTests() {
         testServerMap.clear();
-        runtime.parseConfigurationString(getDefaultConfigurationString())
-                .setCacheDisabled(true); // Disable cache during unit tests to fully stress the system.
```
Comment (kspirov) [inline test/src/test/java/org/corfudb/runtime/view/AbstractViewTest.java]: Forgotten or intentionally?

File: runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java [hunk 7]
```diff
@@ -274,18 +302,38 @@ protected boolean processEntryForContext(final ILogData data) {
...
+    /** Add a new context. */
+    protected void pushNewContext(UUID id, long maxGlobal) {
+        streamContexts.add(contextFactory.apply(id, maxGlobal));
+    }
+
+    protected void popContext() {
+        if (streamContexts.size() <= 1) {
+            throw new RuntimeException("Attempted to pop context with less than 1 context remaining!");
+        }
+        streamContexts.pollFirst();
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java]: Just wandering - why this is FIFO, not stack?

File: runtime/src/main/java/org/corfudb/runtime/view/stream/IStreamView.java [hunk 8]
```diff
@@ -69,6 +112,21 @@ default ILogData next() {
         return nextUpTo(Long.MAX_VALUE);
     }
 
+    /** Retrieve the previous entry in the stream. If there are no previous entries,
+     * NULL will be returned.
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/IStreamView.java]: @javax.annotations.Nullable (for previous(), current() and nextUpTo()).

File: runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java [hunk 9]
```diff
@@ -171,6 +176,96 @@ protected ILogData getNextEntry(QueuedStreamContext context,
...
+        }
+
+        // Otherwise, the previous entry should be resolved, so get
+        // one less than the current.
+        Long prevAddress = context
+                .resolvedQueue.lower(context.globalPointer);
+        // If the pointer is before our min resolution, we need to resolve
+        // to get the correct previous entry.
+        if (prevAddress == null && context.minResolution
+                != Address.NEVER_READ || prevAddress <= context.minResolution) {
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java]: prevAddress can still be null in prevAddress <= context.minResolution

File: runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java [hunk 10]
```diff
@@ -5,6 +5,7 @@
 import org.corfudb.runtime.CorfuRuntime;
 import org.corfudb.runtime.exceptions.OverwriteException;
 import org.corfudb.runtime.view.Address;
+import org.corfudb.util.Utils;
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java]: you don't need it

File: runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java [hunk 11]
```diff
@@ -167,14 +168,25 @@ protected boolean fillReadQueue(final long maxGlobal,
...
             return fillFromResolved(maxGlobal, context);
         }
 
-        // First, we fetch the current token (backpointer) from the sequencer.
-        final long latestToken = runtime.getSequencerView()
-                .nextToken(Collections.singleton(context.id), 0)
-                .getToken();
+        Long latestToken = null;
+
+        // If the max has bveen resolved, use it.
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java]: bveen  -> been

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java [hunk 12]
```diff
@@ -322,69 +230,60 @@ public long commitTransaction() throws TransactionAbortedException {
...
                 .getAddressSpaceView().read(commitAddress);
 
-        if (committedEntry.getType() == DataType.EMPTY) {
-            return;
-        }
-
         updateAllProxies(x -> {
+            log.trace("Commit[{}] Committing {}", this,  x);
+            // Commit all the optimistic updates
+            x.getUnderlyingObject().optimisticCommitUnsafe();
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: the name optimisticCommitUnsafe is legacy, I would change to something like 'resetOptimisticStream'

File: runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java [hunk 13]
```diff
@@ -87,6 +88,25 @@ public synchronized void reset() {
      * {@inheritDoc}
      */
     @Override
+    public synchronized void seek(long globalAddress) {
+        // pop any stream context which has a max address
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java]: This method is public, probably you need to check that (globalAddres>=0) and throw IllegalArgumentException if necessary

File: runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java [hunk 14]
```diff
@@ -87,6 +88,25 @@ public synchronized void reset() {
...
+        while (this.streamContexts.size() > 1) {
+            if (this.streamContexts.first().maxGlobalAddress <
+                    globalAddress)
+            {
+                this.streamContexts.pollFirst();
+            }
+        }
+
+        // now request a seek on the context
+        this.streamContexts.first().seek(globalAddress);
```
Comment (kspirov) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java]: So we are certain it will be single context here, as all other will be removed,  and only one will have maxGlobalAddress=globalAddress, correct?

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **ef0aeda**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #492 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/kwMTrhDpvMEyI2Ds3Jm7RqOPZdeN1oCo)

File: test/src/test/resources/logback-test.xml (hunk 4 above)
Comment (no2chem) [inline test/src/test/resources/logback-test.xml]: added intentionally. Provides some sensible defaults...

File: test/src/test/java/org/corfudb/runtime/view/StreamViewTest.java (hunk 5 above)
Comment (no2chem) [inline test/src/test/java/org/corfudb/runtime/view/StreamViewTest.java]: indeed, and I'll add a new constant

File: runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java (hunk 9 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java]: thanks. good catch!

File: runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java (hunk 3 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java]: No, because you actually need to compare from the root context, and I didn't want to keep/expose the context to the smrstream implementation directly...

File: runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java (hunk 2 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java]: it seems like a copy-and-paste error when the code was restructured. will remove.

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java (hunk 12 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: well, it's actually a merge of the optimistic stream into the smrStream, right?

File: runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java (hunk 13 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java]: ok! thanks

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **5fee699**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #492 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/aYJfM4J3vU7n13RkcFwe91znyJx5XMmS)

File: runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java (hunk 14 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java]: you mean Address.MAX?

**Comment by no2chem** [comment]
Closing this PR to reopen a new clean one after many updates.
